### ISSUE TITLE
Simplify overlap/separation checking functions?

### ISSUE THREAD

**Issue by ianmackenzie** [ianmackenzie/elm-geometry#54]
Switch from

```elm
overlappingBy : Order -> Float -> BoundingBox2d -> BoundingBox2d -> Bool
separatedBy : Order -> Float -> BoundingBox2d -> BoundingBox2d -> Bool
```

to

```elm
overlappingByAtLeast : Float -> BoundingBox2d -> BoundingBox2d -> Bool
separatedByAtLeast : Float -> BoundingBox2d -> BoundingBox2d -> Bool
```

where `overlappingByAtLeast 0` _includes_ touching and `separatedByAtLeast 0` _excludes_ touching.

Several advantages:
  - Simpler (can cut down on the currently-massive  docs!)
  - More efficient (don't have to switch on `Order` value in the implementation, helper functions like `squaredSeparationAmount` can be specialized for what is currently just the `GT` case)
  - Arguably clearer
  - Avoids weird `EQ` case
  - Nicely symmetric with `intersects`: `overlappingByAtLeast 0` means intersecting, `separatedByAtLeast 0` means not intersecting

This wouldn't allow checking for strict intersection, though (overlap of any amount strictly greater than zero). Perhaps `overlappingByMoreThan` and `separatedByLessThan`?

**Comment by ianmackenzie** [comment]
@architectcodes @mpizenberg any thoughts? I'm a bit nervous about reopening the huge discussion we had in #23, but I think there are some good arguments for simplifying things slightly. Tomek, do you still need checking for 'any overlap strictly greater than zero' or would using something like `overlappingByAtLeast 1e-6` work?

**Comment by architectcodes** [comment]
Oh no, for sure `overlappingByAtLeast 1e-6` is more than enough for my needs!

Side note: if I understand things correctly, the nice and simple solution you’re proposing would even cover 'any overlap strictly greater than zero': `not << separatedByAtLeast 0` 🙂

**Comment by mpizenberg** [comment]
For consistency, I would use `>=` for both i.e. returning true with 0. Then in the documentation, mention that one can use `epsilon = 5e-324` which is the smallest number in JS, and thus in elm

**Comment by mpizenberg** [comment]
It also makes sense but I would think that it is more common for people to check that things are separated by at least a certain amount than more than a certain amount. And consequently it would be weird if the 0 case behaved different. But I'm not currently using this though so I don't want to steer you in a wrong direction. I would say probably choose what better fits your current needs (I guess for the delaunay triangulation?) if it works for Tomek also.

**Comment by ianmackenzie** [comment]
I'm not actually using bounding box overlap/separation for anything right now. I think I was just doing some general code/documentation review recently and the complexity of these functions/length of the corresponding documentation was starting to bug me a bit.

How about this as a new plan:

```elm
overlappingByMoreThan : Float -> BoundingBox2d -> BoundingBox2d -> Bool
distanceFrom : BoundingBox2d -> BoundingBox2d -> Float
expandBy : Float -> BoundingBox2d -> BoundingBox2d
```

I think `overlappingByMoreThan` is the clear winner for an overlap predicate. And

```elm
BoundingBox2d.separatedByAtLeast tolerance box1 box2
```

could be replaced with

```elm
not (BoundingBox2d.intersects (BoundingBox2d.expandBy tolerance box1) box2)
```

or

```elm
BoundingBox2d.distanceFrom box1 box2 >= tolerance
```

The first version will be faster for partial application/multiple calls (create an expanded box once and then do lots of cheap intersection tests), while the second will be faster for one-off checks (no new object allocations!) and would correctly return diagonal distances between box corners if that was important. If `distanceFrom` returned a negative value in the overlap case indicating the amount of overlap, then you could even make a single `distanceFrom` call and be able to distinguish between overlapping, touching and separated by just checking the resulting value!

**Comment by architectcodes** [comment]
```elm
BoundingBox2d.distanceFrom box1 box2 >= tolerance
```

Really love the simplicity of it! Perhaps `overlappingByMoreThan` isn’t even necessary in this new world?

**Comment by ianmackenzie** [comment]
> Perhaps `overlappingByMoreThan` isn’t even necessary in this new world?

You're absolutely right that it wouldn't be strictly necessary, but I think

```elm
if BoundingBox2d.overlappingByMoreThan 1e-3 box1 box2 ...
```

is clearer and more readable than

```elm
if BoundingBox2d.distanceFrom box1 box2 < -1e-3 ...
```

and the implementation could be more efficient - `distanceFrom` will need to have some extra branching to check whether the boxes are separated or overlapping, then figure out whether the closest distance is a horizontal/vertical side-to-side or diagonal corner-to-corner distance, do a square root in the case of diagonal distance, etc. Whereas `overlappingByMoreThan` can be implemented with just a couple of subtractions/comparisons. I think `distanceFrom` would be useful primarily if you do need that accurate diagonal distance for some reason, or want to do multi-way branching like

```elm
let
    distance =
        BoundingBox2d.distanceFrom box1 box2
in
if distance > separationTolerance then
    -- boxes are safely separated
else if distance < -overlapTolerance then
    -- boxes overlap badly
else if distance > 0 then
    -- boxes are just barely separated
else
    -- boxes are just touching or slightly overlapping
    
```

**Comment by architectcodes** [comment]
Love it! Thanks for sharing the extra context, that’s very useful! The writeup is so helpful, that perhaps parts of it can be paraphrased in the docs for the new API? Particularly the bits about performance in different scenarios. 💯

**Comment by ianmackenzie** [comment]
TL;DR updated proposal:

```elm
BoundingBox2d.overlappingByMoreThan : Float -> BoundingBox2d -> BoundingBox2d -> Bool
BoundingBox2d.distanceFrom : BoundingBox2d -> BoundingBox2d -> Interval
BoundingBox2d.minDistanceFrom : BoundingBox2d -> BoundingBox2d -> Float
BoundingBox2d.expandBy : Float -> BoundingBox2d -> BoundingBox2d
```

A bit weirdly asymmetric but I think pretty clear/unambiguous and covers most common cases.

---

Thinking about `distanceFrom` a bit more, I wonder if it should return an `Interval` instead, with the semantics being "returns the range of possible distances between a point in the first box and a point in the second box" (which implies the result would never be negative). I think this would be consistent with the various functions proposed in #3 - all of the ones proposed are intervals of signed distances, but I think it could make sense to extend the pattern to things like

```elm
BoundingBox2d.distanceFrom : BoundingBox2d -> BoundingBox2d -> Interval
BoudingBox2d.distanceFromPoint : Point2d -> BoundingBox2d -> Interval
BoundingBox3d.distanceFromAxis : Axis3d -> BoundingBox3d -> Interval
```

(Note that distance from an axis in 2D and distance from a plane in 3D are signed distances, but distance from an axis in 3D is an unsigned distance.) There could also be

```elm
BoundingBox2d.minDistanceFrom : BoundingBox2d -> BoundingBox2d -> Float
BoudingBox2d.minDistanceFromPoint : Point2d -> BoundingBox2d -> Float
BoundingBox3d.minDistanceFromAxis : Axis3d -> BoundingBox3d -> Float
```

which would be equivalent to calling `Interval.minValue` on the result of the `distanceFrom*` functions, but could be a bit more efficient.

This would mean there would no longer be a single number you can check to distinguish between all possible cases, but you can always use a combination of `overlappingByMoreThan`, `intersects` and `minDistanceFrom`. If in the future that becomes a common operation we could add

```elm
module BoundingBox2d

type Classification
    = Touching
    | Overlapping Float
    | Separated Float

classify : BoundingBox2d -> BoundingBox2d -> Classification
```

where the `Float` values in the `Overlapping` and `Separated` cases will always be greater than zero (otherwise the two boxes would be `Touching`!).
